juju/juju
# Switch environ/ec2 "Instances" methods from amz package to aws-sdk-go

This switches the various Environ methods that call EC2's `DescribeInstances` API from the legacy amz package to the aws-sdk-go package. The rationale is that aws-sdk-go has better (more forgiving) retry support, whereas amz only retries on rate-limit responses.

We want to switch to aws-sdk-go for all API calls, but that's too big a chunk of work to bite off for now, so this is my attempt to do it piecemeal, starting for now with instances.

It does mean the mock for `DescribeInstances` is a bit weird. We currently use the amz `ec2test` sub-package to create instances, and the tests rely on those instances being started, so for the `DescribeInstances` mock function to work, we proxy it through to the `Instances` amz call, and convert the request/response types.

## QA steps

To exercise some of the code paths (run `juju debug-log` in another terminal window), first bootstrap with `juju bootstrap aws aws` and check that deploy and open/close ports works. With `firewall-mode=global`:

```
# test on a model with firewall-mode=global
juju add-model fwglobal --config firewall-mode=global
juju model-config firewall-mode  # prints 'global'
juju deploy mediawiki
juju deploy mysql
juju relate mysql mediawiki:db

juju expose mediawiki
# check in AWS console that port 80 is allowed in instance's security group
# also check that http://{ip_address}/ is accessible
# debug log says: opened port ranges [80/tcp from 0.0.0.0/0,::/0] in environment

juju unexpose mediawiki
# port 80 now not allowed in AWS console
# also check that http://{ip_address}/ is not accessible
# debug log says: closed port ranges [80/tcp from 0.0.0.0/0,::/0] in environment

juju destroy-model fwglobal
```

With `firewall-mode=instance` (the default):

```
# test on a model with firewall-mode=instance (the default)
juju add-model fwinstance
juju model-config firewall-mode  # prints 'instance'
juju deploy mediawiki
juju deploy mysql
juju relate mysql mediawiki:db

juju expose mediawiki
# check in AWS console that port 80 is allowed in instance's security group
# also check that http://{ip_address}/ is accessible
# debug log says: opened port ranges [80/tcp from 0.0.0.0/0,::/0] on "machine-0"

juju unexpose mediawiki
# port 80 now not allowed in AWS console
# also check that http://{ip_address}/ is not accessible
# debug log says: closed port ranges [80/tcp from 0.0.0.0/0,::/0] on "machine-0"

juju destroy-model fwinstance
```

Another test I did is to test the auth error handling. I did this by attempting to update the AWS credential on the controller. This calls `checkIAASModelCredential()` which ends up calling `environ.Instances()`, which has the aws-sdk-go error handling. It seems to be handled correctly as a credentials error:

```
$ juju update-credential aws --controller aws
Credential invalid for:
  controller:
    
The provided credentials could not be validated and 
may not be authorized to carry out the request.
Ensure that your account is authorized to use the Amazon EC2 service and 
that you are using the correct access keys. 
These keys are obtained via the "Security Credentials"
page in the AWS console.
: AWS was not able to validate the provided access credentials (AuthFailure)
  default:
    
The provided credentials could not be validated and 
may not be authorized to carry out the request.
Ensure that your account is authorized to use the Amazon EC2 service and 
that you are using the correct access keys. 
These keys are obtained via the "Security Credentials"
page in the AWS console.
: AWS was not able to validate the provided access credentials (AuthFailure)
Failed models may require a different credential.
Use ‘juju set-credential’ to change credential for these models before repeating this update.
```

I also did some local hacking to convince myself that `maybeConvertCredentialError` works the same for both `*ec2.Error` and `awserr.Error`. Tested it by manually calling `env.Instances()`, which calls `groupByName()` using amzec2, then commenting out that so it calls `DescribeInstances` using sdkec2. Both work fine and handle code=AuthFailure correctly.
